sql: remove a couple of allocations in connExecutor.resetPlanner

yuzefovich opened the pull request:
**sql: unify newInternalPlanner and resetPlanner a bit**

This should be a no-op change that removes some of the duplicated
initialization.

**sql: avoid an allocation in connExecutor.resetPlanner**

This commit avoids an allocation for `kvpb.MinTimestampBoundUnsatisfiableError` in `connExecutor.resetPlanner` in the happy path (meaning when there is not auto retry error). This allocation occurred because we're passing a reference to the pointer to the error into `errors.As`, and now we allocate only in the error path.

Additionally, this commit separates out handling of "max timestamp bound" (which is used by the descriptor collection) and locality handling (used by "query home region" feature of the optimizer). In other words, it's not exactly an equivalent change, but I think the previous version was incorrect.

**eval: avoid redundant allocation of UnsupportedTypeChecker**

Given that we currently don't have any types that should be rejected in mixed version clusters, returning `nil` instead of fully-initialized `unsupportedTypeChecker` is equivalent, which allows us to avoid an allocation in `connExecutor.resetPlanner`.

Epic: CRDB-39063.

yuzefovich committed:
sql: avoid an allocation in connExecutor.resetPlanner

This commit avoids an allocation for
`kvpb.MinTimestampBoundUnsatisfiableError` in
`connExecutor.resetPlanner` in the happy path (meaning when there is not
auto retry error). This allocation occurred because we're passing
a reference to the pointer to the error into `errors.As`, and now we
allocate only in the error path.

Additionally, this commit separates out handling of "max timestamp
bound" (which is used by the descriptor collection) and locality
handling (used by "query home region" feature of the optimizer). In
other words, it's not exactly an equivalent change, but I think the
previous version was incorrect.

Release note: None

yuzefovich committed:
eval: avoid redundant allocation of UnsupportedTypeChecker

Given that we currently don't have any types that should be rejected in
mixed version clusters, returning `nil` instead of fully-initialized
`unsupportedTypeChecker` is equivalent, which allows us to avoid an
allocation in `connExecutor.resetPlanner`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125461)
<!-- Reviewable:end -->

yuzefovich committed:
sql: unify newInternalPlanner and resetPlanner a bit

This should be a no-op change that removes some of the duplicated
initialization.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125461#-:-O-4twfe2nlbqqHXs_WN:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->
